### ISSUE TITLE
[PSP2]FMVs freeze on unthrottle/alternate speed

### ISSUE THREAD

**Issue by AkiraJkr** [hrydgard/ppsspp#11207]
It sometimes freezes, sometimes doesn't. But the log is the same for all FMVs.

This is the log when a FMV starts, no difference with unthrottle/alternate speed on. (1.5.4)

![image](https://i.imgur.com/YhnE21l.png)

The same log, no difference again with unthrottle/alternate speed on. (1.6.3)

![image](https://user-images.githubusercontent.com/27028529/41811973-037bf44a-76f0-11e8-93a6-811a6a412b68.png)

**Comment by AkiraJkr** [comment]
It was originally on false, but upon setting to True, things changed soo slightly it was hard to notice, it still kept freezing, but less frequently than before? It is still frequent, and nothing else changed.

Screenshot with FrameskipUnthrottle on True.
![image](https://user-images.githubusercontent.com/27028529/41812546-5b63a7c4-76fb-11e8-91aa-13045c63a486.png)

**Comment by asl97** [comment]
@AkiraJkr When it freezes, does the sound still play?

@unknownbrackets 
On linux, v1.6.3 SDL, I just hit the unthrottle key a few times and the UI freezes, the sound keep playing in fast forward, the memory usage grows very fast.

Pressing close doesn't do anything until the sound plays till the end then it closes.

'Alternate speed' that is not unlimited (0) doesn't seem to cause an issue,  when it's above 100%, the speed seem to get cap at 100%.

The following log is what comes after hitting unthrottle after it finish playing and closes.
```
49:53:762 readThread   I[SCEKERNEL]: HLE/sceKernelThread.cpp:21 sceKernelExitThread(0)
49:54:705 user_main    E[ME]: HLE/sceMpeg.cpp:1256 UNIMPL sceMpegAvcDecodeStopYCbCr(09092700, 090927a4, 09fff3b4)
49:54:707 displayThrea I[SCEKERNEL]: HLE/sceKernelThread.cpp:21 sceKernelExitThread(0)
49:54:708 soundThread  I[SCEKERNEL]: HLE/sceKernelThread.cpp:21 sceKernelExitThread(0)
49:54:710 user_main    I[ME]: HLE/sceMpeg.cpp:1615 sceMpegFinish(...)
49:54:711 user_main    I[SCEUTIL]: HLE/sceUtility.cpp:327 0=sceUtilityUnloadModule(00000303)
49:54:720 idle0        I[FRAMEBUF]: Common/FramebufferCommon.cp Creating FBO for 000cc000 : 480 x 272 x 3
49:54:726 SOUND_THREAD I[ME]: HLE/sceAtrac.cpp:1910 0=sceAtracSetDataAndGetID(09011000, 00080000): atrac3+ stereo audio
49:54:726 SOUND_THREAD E[ME]: HLE/sceAtrac.cpp:1695 80630016=sceAtracResetPlayPosition(0, 0, 0, 0): first byte count not in valid range
51:25:619 user_main    I[FRAMEBUF]: Common/FramebufferCommon.cp Creating FBO for 00044000 : 480 x 272 x 1
^[[A
I: GPU_GLES.cpp:331: GPU_GLES: DeviceLost
51:34:232 user_main    I[G3D]: Common/TextureCacheCommon.cpp:15 Texture cached cleared from 9 textures
51:34:232 user_main    I[FRAMEBUF]: GLES/FramebufferManagerGLES Destroying FBO for 00000000 : 480 x 272 x 1
51:34:232 user_main    I[FRAMEBUF]: GLES/FramebufferManagerGLES Destroying FBO for 00088000 : 480 x 272 x 3
51:34:232 user_main    I[FRAMEBUF]: GLES/FramebufferManagerGLES Destroying FBO for 000cc000 : 480 x 272 x 3
51:34:232 user_main    I[FRAMEBUF]: GLES/FramebufferManagerGLES Destroying FBO for 00044000 : 480 x 272 x 1
I: NativeApp.cpp:746: NativeShutdownGraphics
I: NativeApp.cpp:773: NativeShutdownGraphics done
I: GLRenderManager.cpp:195: GL submission thread paused. Frame=1
I: GLRenderManager.cpp:104: ThreadEnd
^[[A51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 272: Thread "idle0": pc= 08000000 sp= 083fff00  READY     (wt=0 wid=0 wv= 00000000 )
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 273: Thread "idle1": pc= 08000000 sp= 083fef00  READY     (wt=0 wid=0 wv= 00000000 )
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 274: Module "psu": name=psu gp=08e53ca0 entry=08804128
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 276: Thread "user_main": pc= 08bac214 sp= 09fff610 RUN      (wt=0 wid=0 wv= 00000000 )
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 277: MemoryPart "UserSbrk": MemPart: 0909b000 - 090d3000	size: 00038000
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 278: CallBack "EXIT_CB": thread=276, argument= 00000000
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 279: CallBack "POWER_CB": thread=276, argument= 00000000
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 280: Semaphore "FIO_SEMA": -
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 281: Thread "fio_th": pc= 08bb3ad8 sp= 09ff3a90   WAIT    (wt=2 wid=281 wv= 00000000 )
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 282: CallBack "DVDUMD": thread=281, argument= 00000000
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 283: Module "sceFont_Library": faked name=sceFont_Library gp=ffffffff entry=ffffffff
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 284: EventFlag "SceGuSignal": init=00000001 cur=00000001 numwait=0
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 285: FPL "stMem": -
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 286: LwMutex "SceLibsndpSmfExcl": -
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 287: LwMutex "SceLibsndpPrevExcl": -
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 288: Mbx "SOUND MBX": -
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:506 KO 289: Thread "SOUND_THREAD": pc= 08d00c88 sp= 09ff2a90   WAIT    (wt=10 wid=7 wv= 00000400 )
51:39:887 user_main    I[SCEKERNEL]: HLE/sceKernel.cpp:165 Shutting down kernel - 17 kernel objects alive
```

**Comment by AkiraJkr** [comment]
@asl97 Yes, sound plays normally.

**Comment by unknownbrackets** [comment]
Does it ever happen on the intro video (before you even start the game)?

If not, does it happen during the "Somewhere beyond the vastness of space" video?  Or "hurry, the door"?

Trying to understand how to reproduce this most easily.

-[Unknown]

**Comment by asl97** [comment]
I just tap unthrottle on the intro video and it usually freezes before the third press (usually on the first press).
It seem like a timing issue as it doesn't always freeze.

Just keep tapping unthrottle throughout the video and the issue should be reproduced.

**Comment by unknownbrackets** [comment]
Does it only happen in certain backends?  Default settings?

I couldn't reproduce at all, I fast forwarded many times in the middle of the video during about 3 loops.  I tried fast forwarding for a while, fast forwarding until it had already started, fast forwarding 5 short bursts inside, etc.

I also tried both unthrottle and alt speed (unlimited.)

-[Unknown]

**Comment by asl97** [comment]
I only have the opengl backend, I also reproduce it after removing the "~/.config/ppsspp folder" too if that counts as default settings.

I suppose you already tried button mashing the unthrottle button?

Additionally, I notice it doesn't freeze while using "Mode: Skip buffer effects",  also, when switching back to "Mode: Buffered rendering", it doesn't allow unthrottling pass "100% fps" until the end of the video.

**Comment by unknownbrackets** [comment]
Do you have frameskip enabled?

If I use skip buffer effects, the video flickers.  It's smooth with frameskip at 1, though.  So I'm guessing you probably are.

I could definitely see this happening if you use frameskip + are on a platform (like Linux) where we use frameskip during unthrottle.

Ultimately the problem here is that it's registering as 60 FPS during the video, while the video is probably 30 FPS.  So it's drawing nothing for half the frames.  After unthrottling, it can get "on the wrong track" and then only show those nothing frames.

This is similar to, or perhaps the same as, #3417 then.  We need to detect the do-nothing frames better to make sure frameskip never skips ALL the good frames.

-[Unknown]

**Comment by AkiraJkr** [comment]
Yep, do have.

**Comment by asl97** [comment]
@unknownbrackets frameskip is off, video flicker should be normal when it's unbuffered.

I notice FrameSkipUnthrottle was True, setting as false prevent it from unthrottling pass 100% (which prevent the freezing).

So basically you are saying it "gets on the wrong track" and then get stuck processing all those nothing frames without a chance for the UI and other stuff to process, side effect of that is a memory usage explosion.

I wonder why Phantasy Star Portable 2 allows untrottling during the opening while Monster Hunter games doesn't.

**Comment by unknownbrackets** [comment]
Well, if you have frameskip off it should not get stuck, even if FrameSkipUnthrottle is True.  That should at most cause the display not to update while you are unthrottled.

By on the wrong track, I mean if your frameskip setting is 1:

```
GOOD  BAD   GOOD  BAD   GOOD  BAD   GOOD  BAD   GOOD  BAD
 XX          XX          XX          XX          XX
```

That's what we want.  But FrameSkipUnthrottle means we frameskip while you hold down unthrottle, so we could randomly (depending on the timing of your button press) stop one frame later and end up on this "track":

```
GOOD  BAD   GOOD  BAD   GOOD  BAD   GOOD  BAD   GOOD  BAD
       XX          XX          XX          XX          XX
```

That said, if memory usage is exploding, then it seems like you're seeing a different issue.

-[Unknown]

**Comment by asl97** [comment]
One more thing I notice is that if it doesn't freeze when the unthrottle button is press, keeping it hold doesn't cause it to freeze.

It only seem to freeze at the very moment the key is press or is already pressed when the video start.

I tried using valgrind to find the memory leak which should give some kind of hint but it just "Illegal instruction (core dumped)" before the video starts.

Additionally, the memory usage doesn't fall after the video end,  I triggered the issue near the end when testing this so it doesn't use enough memory to stall the system, it still used up close to 1GB from 160MB (before the issue was triggered).

**Comment by unknownbrackets** [comment]
Is this still an issue with the latest git builds after #13865?  This no longer frameskips on unthrottle.

-[Unknown]

**Comment by unknownbrackets** [comment]
Given the change in unthrottling behavior, I'm going to close this.

If you have more information or can confirm it still happens in the latest git builds, please reply to this issue and it'll be reopened.  If you have a new issue with the same game, just create a new issue instead.

-[Unknown]
